File: pkg/video/hls/part.go
package hls

import (
	"bytes"
	"io"
	"nvr/pkg/video/gortsplib"
	"nvr/pkg/video/mp4"
	"nvr/pkg/video/mp4/bitio"
	"strconv"
	"time"
)

type myMdat struct {
	videoSamples []*VideoSample
	audioSamples []*AudioSample
}

func (*myMdat) Type() mp4.BoxType { return mp4.TypeMdat() }

func (b *myMdat) Size() int {
	var total int
	for _, e := range b.videoSamples {
		total += len(e.AVCC)
	}
	for _, e := range b.audioSamples {
		total += len(e.AU)
	}
	return total
}

func (b *myMdat) Marshal(w *bitio.Writer) error {
	for _, e := range b.videoSamples {
		if _, err := w.Write(e.AVCC); err != nil {
			return err
		}
	}
	for _, e := range b.audioSamples {
		if _, err := w.Write(e.AU); err != nil {
			return err
		}
	}
	return nil
}

// NanoToTimescale converts value in nanoseconds into a different timescale.
func NanoToTimescale(value int64, timescale int64) int64 {
	secs := value / int64(time.Second)
	dec := value % int64(time.Second)
	return secs*timescale + dec*timescale/int64(time.Second)
}

func generateVideoTraf( //nolint:funlen
	muxerStartTime int64,
	trackID int,
	videoSamples []*VideoSample,
	dataOffset int32,
) mp4.Boxes {
	/*
	   traf
	   - tfhd
	   - tfdt
	   - trun
	*/

	tfhd := &mp4.Tfhd{
		FullBox: mp4.FullBox{
			Flags: [3]byte{2, 0, 0},
		},
		TrackID: uint32(trackID),
	}

	tfdt := &mp4.Tfdt{
		FullBox: mp4.FullBox{
			Version: 1,
		},
		// sum of decode durations of all earlier samples
		BaseMediaDecodeTimeV1: uint64(
			NanoToTimescale(
				videoSamples[0].DTS-muxerStartTime, VideoTimescale)),
	}

	flags := 0
	flags |= 0x01      // data offset present
	flags |= 0x100     // sample duration present
	flags |= 0x200     // sample size present
	flags |= 0x400     // sample flags present
	flags |= 0x800     // sample composition time offset present or v1
	trun := &mp4.Trun{ // <trun/>
		FullBox: mp4.FullBox{
			Version: 1,
			Flags:   [3]byte{0, byte(flags >> 8), byte(flags)},
		},
		DataOffset: dataOffset,
	}

	trun.Entries = make([]mp4.TrunEntry, len(videoSamples))
	for i, e := range videoSamples {
		off := e.PTS - e.DTS

		flags := uint32(0)
		if !e.IdrPresent {
			flags |= 1 << 16 // sample_is_non_sync_sample
		}
		trun.Entries[i] = mp4.TrunEntry{
			SampleDuration: uint32(
				NanoToTimescale(int64(e.duration()), VideoTimescale)),
			SampleSize:                    uint32(len(e.AVCC)),
			SampleFlags:                   flags,
			SampleCompositionTimeOffsetV1: int32(NanoToTimescale(off, VideoTimescale)),
		}
	}

	return mp4.Boxes{
		Box: &mp4.Traf{},
		Children: []mp4.Boxes{
			{Box: tfhd},
			{Box: tfdt},
			{Box: trun},
		},
	}
}

func generateAudioTraf(
	muxerStartTime int64,
	trackID int,
	audioClockRate int,
	audioSamples []*AudioSample,
	dataOffset int32,
) mp4.Boxes {
	/*
	   traf
	   - tfhd
	   - tfdt
	   - trun
	*/

	tfhd := &mp4.Tfhd{
		FullBox: mp4.FullBox{
			Flags: [3]byte{2, 0, 0},
		},
		TrackID: uint32(trackID),
	}

	tfdt := &mp4.Tfdt{ // <tfdt/>
		FullBox: mp4.FullBox{
			Version: 1,
		},
		BaseMediaDecodeTimeV1: uint64(
			NanoToTimescale(
				audioSamples[0].PTS-muxerStartTime, int64(audioClockRate))),
	}

	flags := 0
	flags |= 0x01  // data offset present
	flags |= 0x100 // sample duration present
	flags |= 0x200 // sample size present

	trun := &mp4.Trun{ // <trun/>
		FullBox: mp4.FullBox{
			Version: 0,
			Flags:   [3]byte{0, byte(flags >> 8), byte(flags)},
		},
		DataOffset: dataOffset,
		Entries:    nil,
	}

	trun.Entries = make([]mp4.TrunEntry, len(audioSamples))
	for i, e := range audioSamples {
		trun.Entries[i] = mp4.TrunEntry{
			SampleDuration: uint32(
				NanoToTimescale(int64(e.Duration()), int64(audioClockRate))),
			SampleSize: uint32(len(e.AU)),
		}
	}

	return mp4.Boxes{
		Box: &mp4.Traf{},
		Children: []mp4.Boxes{
			{Box: tfhd},
			{Box: tfdt},
			{Box: trun},
		},
	}
}

func generatePart( //nolint:funlen
	muxerStartTime int64,
	audioTrack *gortsplib.TrackMPEG4Audio,
	videoSamples []*VideoSample,
	audioSamples []*AudioSample,
) ([]byte, error) {
	/*
	   moof
	   - mfhd
	   - traf (video)
	     - tfhd
	     - tfdt
	     - trun
	   - traf (audio)
	     - tfhd
	     - tfdt
	     - trun
	   mdat
	*/

	moof := mp4.Boxes{
		Box: &mp4.Moof{},
		Children: []mp4.Boxes{
			{Box: &mp4.Mfhd{
				SequenceNumber: 0,
			}},
		},
	}

	mfhdOffset := 24
	videoTrunSize := len(videoSamples)*16 + 20
	audioOffset := mfhdOffset + videoTrunSize + 44

	mdatOffset := audioOffset
	audioTrackExist := audioTrack != nil
	if audioTrackExist && len(audioSamples) != 0 {
		audioTrunOffset := audioOffset + 44
		audioTrunSize := len(audioSamples)*8 + 20
		mdatOffset = audioTrunOffset + audioTrunSize
	}

	trackID := 1
	videoDataOffset := int32(mdatOffset + 8)
	traf := generateVideoTraf(
		muxerStartTime,
		trackID,
		videoSamples,
		videoDataOffset)
	moof.Children = append(moof.Children, traf)
	trackID++

	dataSize := 0
	for _, e := range videoSamples {
		dataSize += len(e.AVCC)
	}
	videoDataSize := dataSize
	if audioTrackExist {
		for _, e := range audioSamples {
			dataSize += len(e.AU)
		}
	}

	if audioTrackExist && len(audioSamples) != 0 {
		audioDataOffset := int32(mdatOffset + 8 + videoDataSize)
		traf := generateAudioTraf(
			muxerStartTime,
			trackID,
			audioTrack.ClockRate(),
			audioSamples,
			audioDataOffset)
		moof.Children = append(moof.Children, traf)
	}

	mdat := &mp4.Boxes{
		Box: &myMdat{
			videoSamples: videoSamples,
			audioSamples: audioSamples,
		},
	}

	size := moof.Size() + mdat.Size()
	buf := bytes.NewBuffer(make([]byte, 0, size))

	w := bitio.NewWriter(buf)

	if err := moof.Marshal(w); err != nil {
		return nil, err
	}

	if err := mdat.Marshal(w); err != nil {
		return nil, err
	}

	return buf.Bytes(), nil
}

func partName(id uint64) string {
	return "part" + strconv.FormatUint(id, 10)
}

// MuxerPart fmp4 part.
type MuxerPart struct {
	audioTrack     *gortsplib.TrackMPEG4Audio
	muxerStartTime int64
	id             uint64

	isIndependent    bool
	VideoSamples     []*VideoSample
	AudioSamples     []*AudioSample
	renderedContent  []byte
	renderedDuration time.Duration
}

func newPart(
	audioTrack *gortsplib.TrackMPEG4Audio,
	muxerStartTime int64,
	id uint64,
) *MuxerPart {
	return &MuxerPart{
		audioTrack:     audioTrack,
		muxerStartTime: muxerStartTime,
		id:             id,
	}
}

func (p *MuxerPart) name() string {
	return partName(p.id)
}

func (p *MuxerPart) reader() io.Reader {
	return bytes.NewReader(p.renderedContent)
}

func (p *MuxerPart) duration() time.Duration {
	total := time.Duration(0)
	for _, e := range p.VideoSamples {
		total += e.duration()
	}
	return total
}

func (p *MuxerPart) finalize() error {
	if len(p.VideoSamples) > 0 || len(p.AudioSamples) > 0 {
		var err error
		p.renderedContent, err = generatePart(
			p.muxerStartTime,
			p.audioTrack,
			p.VideoSamples,
			p.AudioSamples)
		if err != nil {
			return err
		}
		p.renderedDuration = p.duration()
	}

	return nil
}

func (p *MuxerPart) writeH264(sample *VideoSample) {
	if sample.IdrPresent {
		p.isIndependent = true
	}
	p.VideoSamples = append(p.VideoSamples, sample)
}

func (p *MuxerPart) writeAAC(sample *AudioSample) {
	p.AudioSamples = append(p.AudioSamples, sample)
}

File: src/hls/part.rs
use crate::{
    error::{GeneratePartError, GenerateTrafError, PartFinalizeError},
    types::VIDEO_TRACK_ID,
};
use bytes::Bytes;
use common::{
    time::{DurationH264, UnixH264},
    PartFinalized, VideoSample,
};
use mp4::ImmutableBox;
use std::sync::Arc;

fn generate_part(
    muxer_start_time: UnixH264,
    video_samples: Arc<Vec<VideoSample>>,
) -> Result<Bytes, GeneratePartError> {
    /*
       moof
       - mfhd
       - traf (video)
         - tfhd
         - tfdt
         - trun
       mdat
    */

    let mut moof = mp4::Boxes::new(mp4::Moof {}).with_child(
        // Mfhd.
        mp4::Boxes::new(mp4::Mfhd {
            full_box: mp4::FullBox::default(),
            sequence_number: 0,
        }),
    );

    let mfhd_offset = 24;
    let video_trun_size = (video_samples.len() * 16) + 20;
    let mdat_offset = mfhd_offset + video_trun_size + 44;

    let video_data_offset = i32::try_from(mdat_offset + 8)?;
    let traf = generate_traf(muxer_start_time, &video_samples, video_data_offset)?;

    moof.children.push(traf);

    let mdat = mp4::Boxes::new(MyMdat(video_samples));

    let mut buf = Vec::with_capacity(moof.size() + mdat.size());
    moof.marshal(&mut buf)?;
    mdat.marshal(&mut buf)?;

    Ok(Bytes::from(buf))
}

struct MyMdat(Arc<Vec<VideoSample>>);

impl mp4::ImmutableBox for MyMdat {
    fn box_type(&self) -> mp4::BoxType {
        mp4::TYPE_MDAT
    }

    fn size(&self) -> usize {
        self.0.iter().map(|sample| sample.avcc.len()).sum()
    }

    fn marshal(&self, w: &mut dyn std::io::Write) -> Result<(), mp4::Mp4Error> {
        for sample in self.0.iter() {
            w.write_all(&sample.avcc)?;
        }
        Ok(())
    }
}

impl From<MyMdat> for Box<dyn ImmutableBox> {
    fn from(value: MyMdat) -> Self {
        Box::new(value)
    }
}

fn generate_traf(
    muxer_start_time: UnixH264,
    video_samples: &Vec<VideoSample>,
    data_offset: i32,
) -> Result<mp4::Boxes, GenerateTrafError> {
    use GenerateTrafError::*;
    /*
           traf
           - tfhd
           - tfdt
           - trun
    */

    let mut trun_entries = Vec::with_capacity(video_samples.len());
    for sample in video_samples {
        let flags = if sample.random_access_present {
            0
        } else {
            1 << 16 // sample_is_non_sync_sample
        };

        trun_entries.push(mp4::TrunEntry {
            sample_duration: u32::try_from(*sample.duration)
                .map_err(|e| TryFromInt("duration".to_owned(), e))?,
            sample_size: u32::try_from(sample.avcc.len())
                .map_err(|e| TryFromInt("sample_size".to_owned(), e))?,
            sample_flags: flags,
            sample_composition_time_offset_v0: 0,
            sample_composition_time_offset_v1: *sample.dts_offset,
        });
    }

    let first_sample = &video_samples[0];
    let dts = first_sample
        .dts()
        .ok_or(Dts(first_sample.pts, first_sample.dts_offset))?;
    let relative_dts = dts.checked_sub(muxer_start_time).ok_or(Sub)?;
    let base_media_decode_time_v1 = u64::try_from(*relative_dts)
        .map_err(|e| TryFromInt(format!("base_media_decode_time: {relative_dts:?}"), e))?;

    // Traf
    Ok(mp4::Boxes::new(mp4::Traf).with_children3(
        // Tfhd.
        mp4::Boxes::new(mp4::Tfhd {
            full_box: mp4::FullBox {
                version: 0,
                flags: [2, 0, 0],
            },
            track_id: VIDEO_TRACK_ID,
            ..mp4::Tfhd::default()
        }),
        // Tfdt.
        mp4::Boxes::new(mp4::Tfdt {
            full_box: mp4::FullBox {
                version: 1,
                flags: [0, 0, 0],
            },
            // sum of decode durations of all earlier samples
            base_media_decode_time_v0: 0,
            base_media_decode_time_v1,
        }),
        // Trun.
        mp4::Boxes::new(mp4::Trun {
            full_box: mp4::FullBox {
                version: 1,
                flags: mp4::u32_to_flags(
                    mp4::TRUN_DATA_OFFSET_PRESENT
                        | mp4::TRUN_SAMPLE_DURATION_PRESENT
                        | mp4::TRUN_SAMPLE_SIZE_PRESENT
                        | mp4::TRUN_SAMPLE_FLAGS_PRESENT
                        | mp4::TRUN_SAMPLE_COMPOSITION_TIME_OFFSET_PRESENT,
                ),
            },
            data_offset,
            first_sample_flags: 0,
            entries: trun_entries,
        }),
    ))
}

// fmp4 part.
#[derive(Clone)]
#[allow(clippy::module_name_repetitions)]
pub struct MuxerPart {
    pub id: u64,
    pub muxer_start_time: UnixH264,
    pub is_independent: bool,
    pub video_samples: Vec<VideoSample>,
}

impl std::fmt::Debug for MuxerPart {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{} {} {}",
            self.id,
            self.is_independent,
            self.video_samples.len()
        )
    }
}

impl MuxerPart {
    pub fn new(id: u64, muxer_start_time: UnixH264) -> Self {
        Self {
            id,
            muxer_start_time,
            is_independent: false,
            video_samples: Vec::new(),
        }
    }

    pub fn duration(&self) -> Option<DurationH264> {
        let mut total = DurationH264::new(0);
        for e in &self.video_samples {
            total = total.checked_add(e.duration)?;
        }
        Some(total)
    }

    pub fn finalize(self) -> Result<PartFinalized, PartFinalizeError> {
        let rendered_duration = self.duration().ok_or(PartFinalizeError::Duration)?;
        let video_samples = Arc::new(self.video_samples);
        let rendered_content = if video_samples.is_empty() {
            None
        } else {
            Some(generate_part(self.muxer_start_time, video_samples.clone())?)
        };

        Ok(PartFinalized {
            id: self.id,
            is_independent: self.is_independent,
            video_samples: video_samples.clone(),
            rendered_duration,
            rendered_content,
        })
    }

    pub fn write_h264(&mut self, sample: VideoSample) {
        if sample.random_access_present {
            self.is_independent = true;
        }
        self.video_samples.push(sample);
    }
}

#[allow(clippy::unwrap_used)]
#[cfg(test)]
mod tests {
    use super::*;
    use common::time::{DtsOffset, UnixH264, SECOND};
    use pretty_assertions::assert_eq;
    use pretty_hex::pretty_hex;
    use sentryshot_padded_bytes::PaddedBytes;

    #[test]
    fn test_generate_part_minimal() {
        let got = generate_part(UnixH264::new(0), Arc::new(vec![VideoSample::default()])).unwrap();

        let want = vec![
            0, 0, 0, 0x68, b'm', b'o', b'o', b'f', //
            0, 0, 0, 0x10, b'm', b'f', b'h', b'd', //
            0, 0, 0, 0, // FullBox.
            0, 0, 0, 0, // Sequence number.
            0, 0, 0, 0x50, b't', b'r', b'a', b'f', //
            0, 0, 0, 0x10, b't', b'f', b'h', b'd', //
            0, 2, 0, 0, // Track id.
            0, 0, 0, 1, // Sample size.
            0, 0, 0, 0x14, b't', b'f', b'd', b't', //
            1, 0, 0, 0, // Track id.
            0, 0, 0, 0, 0, 0, 0, 0, // BaseMediaDecodeTime.
            0, 0, 0, 0x24, b't', b'r', b'u', b'n', // Video trun.
            1, 0, 0xf, 1, // FullBox.
            0, 0, 0, 1, // Sample count.
            0, 0, 0, 0x70, // Data offset.
            0, 0, 0, 0, // Entry sample duration.
            0, 0, 0, 0, // Entry sample size.
            0, 1, 0, 0, // Entry sample flags.
            0, 0, 0, 0, // Entry SampleCompositionTimeOffset
            0, 0, 0, 8, b'm', b'd', b'a', b't', //
        ];
        assert_eq!(want, got);
    }

    #[test]
    fn test_generate_part_video_sample() {
        let samples = vec![VideoSample {
            avcc: Arc::new(PaddedBytes::new(b"abcd".to_vec())),
            ..Default::default()
        }];

        let got = generate_part(UnixH264::new(0), Arc::new(samples)).unwrap();
        let want = vec![
            0, 0, 0, 0x68, b'm', b'o', b'o', b'f', //
            0, 0, 0, 0x10, b'm', b'f', b'h', b'd', //
            0, 0, 0, 0, // FullBox.
            0, 0, 0, 0, // Sequence number.
            0, 0, 0, 0x50, b't', b'r', b'a', b'f', // Video traf.
            0, 0, 0, 0x10, b't', b'f', b'h', b'd', // Video tfhd.
            0, 2, 0, 0, // Track id.
            0, 0, 0, 1, // Sample size.
            0, 0, 0, 0x14, b't', b'f', b'd', b't', // Video tfdt.
            1, 0, 0, 0, // Track id.
            0, 0, 0, 0, 0, 0, 0, 0, // BaseMediaDecodeTime.
            0, 0, 0, 0x24, b't', b'r', b'u', b'n', // Video trun.
            1, 0, 0xf, 1, // FullBox.
            0, 0, 0, 1, // Sample count.
            0, 0, 0, 0x70, // Data offset.
            0, 0, 0, 0, // Entry sample duration.
            0, 0, 0, 4, // Entry sample size.
            0, 1, 0, 0, // Entry sample flags.
            0, 0, 0, 0, // Entry SampleCompositionTimeOffset
            0, 0, 0, 0xc, b'm', b'd', b'a', b't', //
            b'a', b'b', b'c', b'd', // Video Sample
        ];
        assert_eq!(want, got);
    }

    #[test]
    fn test_generate_part_multiple_video_samples() {
        let samples = vec![
            VideoSample {
                avcc: Arc::new(PaddedBytes::new(b"abcd".to_vec())),
                random_access_present: true,
                ..Default::default()
            },
            VideoSample {
                avcc: Arc::new(PaddedBytes::new(b"efgh".to_vec())),
                ..Default::default()
            },
            VideoSample {
                avcc: Arc::new(PaddedBytes::new(b"ijkl".to_vec())),
                ..Default::default()
            },
        ];

        let got = generate_part(UnixH264::new(0), Arc::new(samples)).unwrap();

        let want = vec![
            0, 0, 0, 0x88, b'm', b'o', b'o', b'f', //
            0, 0, 0, 0x10, b'm', b'f', b'h', b'd', //
            0, 0, 0, 0, // FullBox.
            0, 0, 0, 0, // Sequence number.
            0, 0, 0, 0x70, b't', b'r', b'a', b'f', // Video traf.
            0, 0, 0, 0x10, b't', b'f', b'h', b'd', // Video tfhd.
            0, 2, 0, 0, // Track id.
            0, 0, 0, 1, // Sample size.
            0, 0, 0, 0x14, b't', b'f', b'd', b't', // Video tfdt.
            1, 0, 0, 0, // Track id.
            0, 0, 0, 0, 0, 0, 0, 0, // BaseMediaDecodeTime.
            0, 0, 0, 0x44, b't', b'r', b'u', b'n', // Video trun.
            1, 0, 0xf, 1, // FullBox.
            0, 0, 0, 3, // Sample count.
            0, 0, 0, 0x90, // Data offset.
            0, 0, 0, 0, // Entry1 sample duration.
            0, 0, 0, 4, // Entry1 sample size.
            0, 0, 0, 0, // Entry1 sample flags.
            0, 0, 0, 0, // Entry1 SampleCompositionTimeOffset
            0, 0, 0, 0, // Entry2 sample duration.
            0, 0, 0, 4, // Entry2 sample size.
            0, 1, 0, 0, // Entry2 sample flags.
            0, 0, 0, 0, // Entry2 SampleCompositionTimeOffset
            0, 0, 0, 0, // Entry3 sample duration.
            0, 0, 0, 4, // Entry3 sample size.
            0, 1, 0, 0, // Entry3 sample flags.
            0, 0, 0, 0, // Entry3 SampleCompositionTimeOffset
            0, 0, 0, 0x14, b'm', b'd', b'a', b't', //
            b'a', b'b', b'c', b'd', b'e', b'f', b'g', b'h', b'i', b'j', b'k',
            b'l', // Video Samples
        ];
        assert_eq!(want, got);
    }

    #[test]
    fn test_generate_part_minimal_real() {
        let start_time = UnixH264::new(1_000_000_000 * SECOND);
        let samples = vec![
            VideoSample {
                pts: start_time + UnixH264::new(54000),
                dts_offset: DtsOffset::new(54000 - 60000),
                avcc: Arc::new(PaddedBytes::new(b"abcd".to_vec())),
                random_access_present: true,
                duration: DurationH264::new(11999),
            },
            VideoSample {
                pts: start_time + UnixH264::new(63000),
                dts_offset: DtsOffset::new(63000 - 72000),
                avcc: Arc::new(PaddedBytes::new(b"efgh".to_vec())),
                random_access_present: false,
                duration: DurationH264::new(9000),
            },
        ];

        let got = generate_part(start_time, Arc::new(samples)).unwrap();

        let want = vec![
            0, 0, 0, 0x78, b'm', b'o', b'o', b'f', //
            0, 0, 0, 0x10, b'm', b'f', b'h', b'd', //
            0, 0, 0, 0, // FullBox.
            0, 0, 0, 0, // Sequence number.
            0, 0, 0, 0x60, b't', b'r', b'a', b'f', // Video traf.
            0, 0, 0, 0x10, b't', b'f', b'h', b'd', // Video tfhd.
            0, 2, 0, 0, // Track id.
            0, 0, 0, 1, // Sample size.
            0, 0, 0, 0x14, b't', b'f', b'd', b't', // Video tfdt.
            1, 0, 0, 0, // Track id.
            0, 0, 0, 0, 0, 0, 0xea, 0x60, // BaseMediaDecodeTime.
            0, 0, 0, 0x34, b't', b'r', b'u', b'n', // Video trun.
            1, 0, 0xf, 1, // FullBox.
            0, 0, 0, 2, // Sample count.
            0, 0, 0, 0x80, // Data offset.
            0, 0, 0x2e, 0xdf, // Entry1 sample duration.
            0, 0, 0, 4, // Entry1 sample size.
            0, 0, 0, 0, // Entry1 sample flags.
            0xff, 0xff, 0xe8, 0x90, // 1 Entry SampleCompositionTimeOffset
            0, 0, 0x23, 0x28, // 2 Entry sample duration.
            0, 0, 0, 4, // 2 Entry sample size.
            0, 1, 0, 0, // 2 Entry sample flags.
            0xff, 0xff, 0xdc, 0xd8, // Entry2 SampleCompositionTimeOffset
            0, 0, 0, 0x10, b'm', b'd', b'a', b't', //
            b'a', b'b', b'c', b'd', b'e', b'f', b'g', b'h', // Samples
        ];
        assert_eq!(pretty_hex(&want), pretty_hex(&got));
    }
}
